src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,302 @@
+    /// the map is a <see cref="ConcurrentDictionary{TKey, TValue}"/> which may be modified directly inside an API call or
+    /// through eventual processing of the event queue. This implementation relies on the principle of 'eventual consistency':
+    /// though the map and it's corresponding linked list may be out of sync at any given point in time, they will eventually line up.
+    /// </summary>
+    /// <typeparam name="TKey">The key type to be used by the cache.</typeparam>
+    /// <typeparam name="TValue">The value type to be used by the cache</typeparam>
+    internal class EventBasedLRUCache<TKey, TValue> : ILRUCache<TKey,TValue>, IDisposable
+    {
+        private readonly int _capacity;
+        private readonly ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>> _map;
+        private readonly LinkedList<LRUCacheItem<TKey, TValue>> _doubleLinkedList = new LinkedList<LRUCacheItem<TKey, TValue>>();
+        private readonly BlockingCollection<Action> _eventQueue = new BlockingCollection<Action>();
+        private bool _disposed = false;
+
+        internal EventBasedLRUCache(int capacity, IEqualityComparer<TKey> comparer = null)
+        {
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            new Task(() => OnStart(), TaskCreationOptions.LongRunning).Start();
+            _ = RemoveExpiredValuesPeriodically(TimeSpan.FromMinutes(5));
+        }
+
+        private void OnStart()
+        {
+            while (true)
+            {
+                _eventQueue.Take().Invoke();
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        async Task RemoveExpiredValuesPeriodically(TimeSpan interval)
+        {
+            while (true)
+            {
+                _eventQueue.Add(() => RemoveExpiredValues());
+                await Task.Delay(interval).ConfigureAwait(false);
+            }
+        }
+
+        public void SetValue(TKey key, TValue value)
+        {
+            SetValue(key, value, DateTime.MaxValue);
```

Comment:
The cache works the way it does now because this is the approach used by both the System.Runtime.Caching and Microsoft.Extensions.Caching.Memory libraries. Because of this, I was operating under the assumption that InMemoryCryptoProviderCache would be responsible for setting this value.

That being said, I can see the argument for this given our use case. I'm open to modifying it if we agree using a finite default value makes more sense.